File: cpp/src/lib/SemiSemiParselet.cpp

#include "Parselet.h"

#include "ParseletRegistration.h" // for prefixParselets
#include "ParserSession.h"
#include "SymbolRegistration.h"
#include "Parser.h"
#include "Tokenizer.h"
#include "TokenEnumRegistration.h"

#if USE_MUSTTAIL
#define MUSTTAIL [[clang::musttail]]
#else
#define MUSTTAIL
#endif // USE_MUSTTAIL

//
// SemiSemiParselet is complicated enough to warrant its own implementation file.
// The syntax for ;; is complicated and has a lot of edge cases.
//

SemiSemiParselet::SemiSemiParselet() : PrefixParselet(SemiSemiParselet_parsePrefix), InfixParselet(SemiSemiParselet_parseInfix) {}

Precedence SemiSemiParselet::getPrecedence(ParserSessionPtr session) const {
    return PRECEDENCE_SEMISEMI;
}

Token SemiSemiParselet::processImplicitTimes(ParserSessionPtr session, Token TokIn) const {
    
    //
    // SemiSemi was already parsed with look-ahead with the assumption that implicit Times will be handled correctly
    //
    
    if (Parser_checkSpan(session)) {
        return Token(TOKEN_FAKE_IMPLICITTIMES, TokIn.Buf, 0, Source(TokIn.Src.Start));
    }
    
    return TokIn;
}

void SemiSemiParselet_parsePrefix(ParserSessionPtr session, ParseletPtr Ignored, Token TokIn) {
    
#if CHECK_ABORT
    if (session->abortQ()) {
        Parser_pushNode(session, new AbortNode());
        return Parser_tryContinue(session, Ignored, TokIn/*ignored*/);
    }
#endif // CHECK_ABORT
    
    Parser_pushLeaf(session, Token(TOKEN_FAKE_IMPLICITONE, TokIn.Buf, 0, Source(TokIn.Src.Start)));
    
    Parser_pushContext(session, PRECEDENCE_SEMISEMI);
    
    //
    // nextToken() is not needed after an implicit token
    //
    
    MUSTTAIL
    return SemiSemiParselet_parseInfix(session, Ignored, TokIn);
}

void SemiSemiParselet_parseInfix(ParserSessionPtr session, ParseletPtr Ignored, Token TokIn) {
    
#if CHECK_ABORT
    if (session->abortQ()) {
        Parser_popContext(session);
        Parser_pushNode(session, new AbortNode());
        return Parser_tryContinue(session, Ignored, TokIn/*ignored*/);
    }
#endif // CHECK_ABORT
    
    Parser_pushLeafAndNext(session, TokIn);
    
    MUSTTAIL
    return SemiSemiParselet_parse1(session, Ignored, TokIn/*ignored*/);
}

void SemiSemiParselet_parse1(ParserSessionPtr session, ParseletPtr Ignored, Token Ignored2) {
    
#if CHECK_ABORT
    if (session->abortQ()) {
        Parser_popContext(session);
        Parser_pushNode(session, new AbortNode());
        return Parser_tryContinue(session, Ignored, Ignored2);
    }
#endif // CHECK_ABORT
    
    auto SecondTok = Tokenizer_currentToken(session, TOPLEVEL);
    
    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines(session, SecondTok, TOPLEVEL);
    
    //
    // a;;
    //  ^~TokIn
    //
    
    if (!SecondTok.Tok.isPossibleBeginning()) {
        
        //
        // a;;&
        //    ^SecondTok
        //
        
        Parser_pushLeaf(session, Token(TOKEN_FAKE_IMPLICITALL, SecondTok.Buf, 0, Source(SecondTok.Src.Start)));
        
        //
        // nextToken() is not needed after an implicit token
        //
        
        MUSTTAIL
        return SemiSemiParselet_reduceBinary(session, Ignored, Ignored2);
    }
    
    if (SecondTok.Tok != TOKEN_SEMISEMI) {
        
        //
        // a;;b
        //    ^SecondTok
        //
        
        auto& Ctxt = Parser_topContext(session);
        assert(!Ctxt.F);
        Ctxt.F = SemiSemiParselet_parse2;
        
        auto P2 = prefixParselets[SecondTok.Tok.value()];
        
        MUSTTAIL
        return (P2->parsePrefix)(session, P2, SecondTok);
    }
    
    //
    // a;;;;
    //    ^~SecondTok
    //
    
    Parser_pushLeaf(session, Token(TOKEN_FAKE_IMPLICITALL, SecondTok.Buf, 0, Source(SecondTok.Src.Start)));
    
    SecondTok.skip(session);
    
    auto ThirdTok = Tokenizer_currentToken(session, TOPLEVEL);
    
    auto& Trivia1 = session->trivia1;
    
    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines(session, ThirdTok, TOPLEVEL, Trivia1);
    
    if (!ThirdTok.Tok.isPossibleBeginning() || ThirdTok.Tok == TOKEN_SEMISEMI) {
        
        //
        // a;;;;&
        //      ^ThirdTok
        //
        
        //
        // a;;;;;;
        //      ^~ThirdTok
        //
        
        Trivia1.reset(session);
        SecondTok.reset(session);
        
        MUSTTAIL
        return SemiSemiParselet_reduceBinary(session, Ignored, Ignored2);
    }
    
    //
    // a;;;;b
    //      ^ThirdTok
    //
    
    Parser_pushLeaf(session, SecondTok);
    
    //
    // nextToken() already handled above
    //
    
    Parser_pushTriviaSeq(session, Trivia1);
    
    auto& Ctxt = Parser_topContext(session);
    assert(!Ctxt.F);
    Ctxt.F = SemiSemiParselet_reduceTernary;
    
    auto P2 = prefixParselets[ThirdTok.Tok.value()];
    
    MUSTTAIL
    return (P2->parsePrefix)(session, P2, ThirdTok);
}

void SemiSemiParselet_parse2(ParserSessionPtr session, ParseletPtr Ignored, Token Ignored2) {
    
#if CHECK_ABORT
    if (session->abortQ()) {
        Parser_popNode(session);
        Parser_popContext(session);
        Parser_pushNode(session, new AbortNode());
        return Parser_tryContinue(session, Ignored, Ignored2);
    }
#endif // CHECK_ABORT
    
    auto& Trivia1 = session->trivia1;
    
    auto ThirdTok = Tokenizer_currentToken(session, TOPLEVEL);
    
    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines(session, ThirdTok, TOPLEVEL, Trivia1);
    
    if (!ThirdTok.Tok.isPossibleBeginning() || ThirdTok.Tok != TOKEN_SEMISEMI) {
        
        //
        // a;;b&
        //     ^ThirdTok
        //
        
        //
        // \[Integral];;x\[DifferentialD]x
        //               ^~~~~~~~~~~~~~~~ThirdTok
        //
        
        Trivia1.reset(session);
        
        MUSTTAIL
        return SemiSemiParselet_reduceBinary(session, Ignored, Ignored2);
    }
    
    //
    // a;;b;;
    //     ^~ThirdTok
    //
    
    ThirdTok.skip(session);
    
    auto& Trivia2 = session->trivia2;
    
    auto FourthTok = Tokenizer_currentToken(session, TOPLEVEL);
    
    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines(session, FourthTok, TOPLEVEL, Trivia2);
    
    if (!FourthTok.Tok.isPossibleBeginning() || FourthTok.Tok == TOKEN_SEMISEMI) {
        
        //
        // a;;b;;&
        //       ^FourthTok
        //
        
        //
        // a;;b;;;;
        //       ^~FourthTok
        //
        
        Trivia2.reset(session);
        ThirdTok.reset(session);
        Trivia1.reset(session);
        
        MUSTTAIL
        return SemiSemiParselet_reduceBinary(session, Ignored, Ignored2);
    }
        
    //
    // a;;b;;c
    //       ^FourthTok
    //
    
    Parser_pushTriviaSeq(session, Trivia1);
    
    Parser_pushLeaf(session, ThirdTok);
    
    //
    // nextToken() already handled above
    //
    
    Parser_pushTriviaSeq(session, Trivia2);
    
    auto& Ctxt = Parser_topContext(session);
    assert(Ctxt.F == SemiSemiParselet_parse2);
    Ctxt.F = SemiSemiParselet_reduceTernary;
    
    auto P2 = prefixParselets[FourthTok.Tok.value()];
    
    MUSTTAIL
    return (P2->parsePrefix)(session, P2, FourthTok);
}

void SemiSemiParselet_reduceBinary(ParserSessionPtr session, ParseletPtr Ignored, Token Ignored2) {
    
    Parser_pushNode(session, new BinaryNode(SYMBOL_SPAN, Parser_popContext(session)));
    
    MUSTTAIL
    return Parser_parseClimb(session, Ignored, Ignored2);
}

void SemiSemiParselet_reduceTernary(ParserSessionPtr session, ParseletPtr Ignored, Token Ignored2) {
    
    Parser_pushNode(session, new TernaryNode(SYMBOL_SPAN, Parser_popContext(session)));
    
    MUSTTAIL
    return Parser_parseClimb(session, Ignored, Ignored2);
}

File: crate/src/semi_semi_parselet.rs
use crate::{
    node::{BinaryNode, TernaryNode},
    panic_if_aborted,
    parselet::*,
    parser::{
        Parser_checkSpan, Parser_eatTriviaButNotToplevelNewlines,
        Parser_eatTriviaButNotToplevelNewlines_2, Parser_parseClimb, Parser_popContext,
        Parser_pushContext, Parser_pushLeaf, Parser_pushLeafAndNext, Parser_pushNode,
        Parser_pushTriviaSeq, Parser_topContext,
    },
    parser_session::ParserSession,
    precedence::*,
    source::Source,
    source::TOPLEVEL,
    symbol_registration::SYMBOL_SPAN,
    token::Token,
    token_enum_registration::TokenEnum::*,
    tokenizer::Tokenizer_currentToken,
};

//
// SemiSemiParselet is complicated enough to warrant its own implementation file.
// The syntax for ;; is complicated and has a lot of edge cases.
//

impl InfixParselet for SemiSemiParselet {
    fn parse_infix(&'static self, session: &mut ParserSession, token: Token) {
        SemiSemiParselet_parseInfix(session, token)
    }

    fn getPrecedence(&self, _: &mut ParserSession) -> Precedence {
        return PRECEDENCE_SEMISEMI;
    }

    fn processImplicitTimes(&self, session: &mut ParserSession, TokIn: Token) -> Token {
        //
        // SemiSemi was already parsed with look-ahead with the assumption that implicit Times will be handled correctly
        //

        if Parser_checkSpan(session) {
            return Token::new2(
                TOKEN_FAKE_IMPLICITTIMES,
                TokIn.span,
                Source::from_location(TokIn.src.start),
            );
        }

        return TokIn;
    }
}

impl PrefixParselet for SemiSemiParselet {
    fn parse_prefix(&'static self, session: &mut ParserSession, token: Token) {
        SemiSemiParselet_parsePrefix(session, token)
    }
}

fn SemiSemiParselet_parsePrefix(session: &mut ParserSession, TokIn: Token) {
    panic_if_aborted!();


    Parser_pushLeaf(
        session,
        Token::new2(
            TOKEN_FAKE_IMPLICITONE,
            TokIn.span,
            Source::from_location(TokIn.src.start),
        ),
    );

    Parser_pushContext(session, PRECEDENCE_SEMISEMI);

    //
    // nextToken() is not needed after an implicit token
    //

    // MUSTTAIL
    return SemiSemiParselet_parseInfix(session, TokIn);
}

fn SemiSemiParselet_parseInfix(session: &mut ParserSession, TokIn: Token) {
    panic_if_aborted!();


    Parser_pushLeafAndNext(session, TokIn);

    // MUSTTAIL
    return SemiSemiParselet_parse1(session);
}

fn SemiSemiParselet_parse1(session: &mut ParserSession) {
    panic_if_aborted!();


    let mut SecondTok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines(session, &mut SecondTok, TOPLEVEL);

    //
    // a;;
    //  ^~TokIn
    //

    if !SecondTok.tok.isPossibleBeginning() {
        //
        // a;;&
        //    ^SecondTok
        //

        Parser_pushLeaf(
            session,
            Token::new2(
                TOKEN_FAKE_IMPLICITALL,
                SecondTok.span,
                Source::from_location(SecondTok.src.start),
            ),
        );

        //
        // nextToken() is not needed after an implicit token
        //

        // MUSTTAIL
        return SemiSemiParselet_reduceBinary(session);
    }

    if SecondTok.tok != TOKEN_SEMISEMI {
        //
        // a;;b
        //    ^SecondTok
        //

        let Ctxt = Parser_topContext(session);
        assert!(Ctxt.f.is_none());
        Ctxt.f = Some(|s, _| SemiSemiParselet_parse2(s));

        let P2 = prefix_parselet(SecondTok.tok);

        // MUSTTAIL
        return P2.parse_prefix(session, SecondTok);
    }

    //
    // a;;;;
    //    ^~SecondTok
    //

    Parser_pushLeaf(
        session,
        Token::new2(
            TOKEN_FAKE_IMPLICITALL,
            SecondTok.span,
            Source::from_location(SecondTok.src.start),
        ),
    );

    SecondTok.skip(&mut session.tokenizer);

    let mut ThirdTok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    let Trivia1 = session.trivia1.clone();

    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines_2(
        session,
        &mut ThirdTok,
        TOPLEVEL,
        &mut Trivia1.borrow_mut(),
    );

    if !ThirdTok.tok.isPossibleBeginning() || ThirdTok.tok == TOKEN_SEMISEMI {
        //
        // a;;;;&
        //      ^ThirdTok
        //

        //
        // a;;;;;;
        //      ^~ThirdTok
        //

        Trivia1.borrow_mut().reset(&mut session.tokenizer);
        SecondTok.reset(&mut session.tokenizer);

        // MUSTTAIL
        return SemiSemiParselet_reduceBinary(session);
    }

    //
    // a;;;;b
    //      ^ThirdTok
    //

    Parser_pushLeaf(session, SecondTok);

    //
    // nextToken() already handled above
    //

    Parser_pushTriviaSeq(session, &mut Trivia1.borrow_mut());

    let Ctxt = Parser_topContext(session);
    assert!(Ctxt.f.is_none());
    Ctxt.f = Some(|s, _| SemiSemiParselet_reduceTernary(s));

    let P2 = prefix_parselet(ThirdTok.tok);

    // MUSTTAIL
    return P2.parse_prefix(session, ThirdTok);
}

fn SemiSemiParselet_parse2(session: &mut ParserSession) {
    panic_if_aborted!();


    let Trivia1 = session.trivia1.clone();

    let mut ThirdTok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines_2(
        session,
        &mut ThirdTok,
        TOPLEVEL,
        &mut Trivia1.borrow_mut(),
    );

    if !ThirdTok.tok.isPossibleBeginning() || ThirdTok.tok != TOKEN_SEMISEMI {
        //
        // a;;b&
        //     ^ThirdTok
        //

        //
        // \[Integral];;x\[DifferentialD]x
        //               ^~~~~~~~~~~~~~~~ThirdTok
        //

        Trivia1.borrow_mut().reset(&mut session.tokenizer);

        // MUSTTAIL
        return SemiSemiParselet_reduceBinary(session);
    }

    //
    // a;;b;;
    //     ^~ThirdTok
    //

    ThirdTok.skip(&mut session.tokenizer);

    let Trivia2 = session.trivia2.clone();

    let mut FourthTok = Tokenizer_currentToken(&mut session.tokenizer, TOPLEVEL);

    //
    // Span should not cross toplevel newlines
    //
    Parser_eatTriviaButNotToplevelNewlines_2(
        session,
        &mut FourthTok,
        TOPLEVEL,
        &mut Trivia2.borrow_mut(),
    );

    if !FourthTok.tok.isPossibleBeginning() || FourthTok.tok == TOKEN_SEMISEMI {
        //
        // a;;b;;&
        //       ^FourthTok
        //

        //
        // a;;b;;;;
        //       ^~FourthTok
        //

        Trivia2.borrow_mut().reset(&mut session.tokenizer);
        ThirdTok.reset(&mut session.tokenizer);
        Trivia1.borrow_mut().reset(&mut session.tokenizer);

        // MUSTTAIL
        return SemiSemiParselet_reduceBinary(session);
    }

    //
    // a;;b;;c
    //       ^FourthTok
    //

    Parser_pushTriviaSeq(session, &mut Trivia1.borrow_mut());

    Parser_pushLeaf(session, ThirdTok);

    //
    // nextToken() already handled above
    //

    Parser_pushTriviaSeq(session, &mut Trivia2.borrow_mut());

    let Ctxt = Parser_topContext(session);
    assert!(Ctxt.f.unwrap() as usize == SemiSemiParselet_parse2 as usize);
    Ctxt.f = Some(|s, _| SemiSemiParselet_reduceTernary(s));

    let P2 = prefix_parselet(FourthTok.tok);

    // MUSTTAIL
    return P2.parse_prefix(session, FourthTok);
}

fn SemiSemiParselet_reduceBinary(session: &mut ParserSession) {
    let node = BinaryNode::new(SYMBOL_SPAN, Parser_popContext(session));
    Parser_pushNode(session, node);

    // MUSTTAIL
    return Parser_parseClimb(session);
}

fn SemiSemiParselet_reduceTernary(session: &mut ParserSession) {
    let node = TernaryNode::new(SYMBOL_SPAN, Parser_popContext(session));
    Parser_pushNode(session, node);

    // MUSTTAIL
    return Parser_parseClimb(session);
}
